File: sdks/java/src/test/java/org/byteveda/flexiq/core/NamespaceScopingTest.java
package org.byteveda.flexiq.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Path;
import org.byteveda.flexiq.FlexiQ;
import org.byteveda.flexiq.task.Task;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.io.TempDir;

/**
 * A namespace is a tenancy boundary: a caller scoped to one must learn nothing
 * about ids outside it — not through a read, and not through the effect of a
 * write. An unset namespace stays unscoped and addresses every namespace.
 */
class NamespaceScopingTest {

    private static FlexiQ open(Path db, String namespace) {
        FlexiQ.Builder builder = FlexiQ.builder().backend("sqlite").url(db.toString());
        if (namespace != null) {
            builder = builder.namespace(namespace);
        }
        return builder.open();
    }

    @Test
    @Timeout(30)
    void scopesTheIdAddressedSurface(@TempDir Path dir) {
        Path db = dir.resolve("t.db");
        Task<String> work = Task.of("work", String.class);

        try (FlexiQ a = open(db, "ns-a");
                FlexiQ b = open(db, "ns-b");
                FlexiQ unscoped = open(db, null)) {
            String id = a.enqueue(work, "payload");

            assertTrue(a.getJob(id).isPresent(), "the owning namespace sees its job");
            assertTrue(b.getJob(id).isEmpty(), "another namespace reads it as missing");
            assertTrue(unscoped.getJob(id).isPresent(), "unscoped addresses every namespace");

            assertFalse(b.cancel(id), "a cross-namespace cancel reports like an unknown id");
            assertFalse(b.requestCancel(id));
            assertTrue(a.getJob(id).isPresent(), "and must not have landed");

            assertEquals(0, b.jobErrors(id).size());
            assertEquals(0, b.getTaskLogs(id).size());

            assertTrue(a.cancel(id), "the owning namespace still cancels");
        }
    }
}

File: sdks/node/test/core/namespaceScoping.test.ts
// A namespace is a tenancy boundary: a caller scoped to one must learn nothing
// about ids outside it — not through a read, and not through the effect of a
// write. `undefined` stays unscoped and addresses every namespace.

import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { expect, it } from "vitest";
import { Queue } from "../../src/index";

function sharedDb(): string {
  return join(mkdtempSync(join(tmpdir(), "flexiq-ns-")), "queue.db");
}

it("scopes the id-addressed surface to the queue's namespace", async () => {
  const dbPath = sharedDb();
  const a = new Queue({ dbPath, namespace: "ns-a" });
  const b = new Queue({ dbPath, namespace: "ns-b" });
  const unscoped = new Queue({ dbPath });

  a.task("work", () => undefined);
  const id = a.enqueue("work", []);

  expect(a.getJob(id)).not.toBeNull();
  expect(b.getJob(id)).toBeNull();
  expect(unscoped.getJob(id)).not.toBeNull();

  expect(b.cancelJob(id)).toBe(false);
  expect(b.requestCancel(id)).toBe(false);
  expect(a.getJob(id)).not.toBeNull();

  expect(await b.getJobErrors(id)).toEqual([]);
  expect(b.taskLogs(id)).toEqual([]);

  expect(a.cancelJob(id)).toBe(true);
});

it("scopes the archive listing", async () => {
  const dbPath = sharedDb();
  const a = new Queue({ dbPath, namespace: "ns-a" });
  const b = new Queue({ dbPath, namespace: "ns-b" });
  const unscoped = new Queue({ dbPath });

  for (const queue of [a, b]) {
    queue.task("work", () => undefined);
    queue.cancelJob(queue.enqueue("work", []));
  }

  expect(await a.listArchived()).toHaveLength(1);
  expect((await a.listArchived())[0]?.namespace).toBe("ns-a");
  expect(await unscoped.listArchived()).toHaveLength(2);
  expect((await a.listArchivedAfter()).items).toHaveLength(1);
});
